### PR TITLE
WIP: VTOL roll trim 

### PR DESCRIPTION
**Pull request by sanderux**
This adds a roll trim param to combat roll produced by the inertia of the pusher prop.
Also, the scaling was removed for roll and pitch during transition in order to correctly apply these values.

With this param, and the already existing pitch trim the climbing and rolling during transition can be corrected. The roll trim will have to be set depending on the rotation direction of the pusher / puller. In SITL the following values produce a stable result;
VT_FW_PITCH_TRIM: -0.25
VT_FW_ROLL_TRIM: 0.25

The MC, during transition blending does not have enough authority to correct roll and pitch and the FW is not yet in control to apply these corrections.

Setting the above values fixes https://github.com/PX4/Firmware/issues/4079
Airframes should independently be analyzed for correct values.
Roll should be set positive CCW pusher prop and negative with a CW pusher prop
Pitch (in all the logs i have seen) always seems to angle up and therefor needs a negative value

I'm not sure if this is a permanent solution or if we should revisit the control authority of MC/FW during transitions.

@AndreasAntener @tumbili @LorenzMeier

### CODE REVIEW

**Comment by tubeme** [comment]
Sorry for the interruption, but in my opinion controlling the aircraft with fixed trims during flight for level transition or just flight, is totally in contrast with having so many sensors and ways to stabilize the machine during all phases of flight including transition. In my opinion this transition becomes a behemoth already with so many aspects to twitch and set so many combinations of parameters that at some point it will become unmanageable.

We have a VERY VERY simple task to do. From Stabilized flight MC we go in stabilized Forward Flight. During the acceleration phase (until lift is produced) is the quad is capable to give perfect stabilization! Then the airspeed trigers clean FW flight once we have lift, and from that point on we have stabilization done by the plane.

We need simple solution for simple things. This functionality is basic functionality it should be fast robust and clean.

As I mentioned before we created KK2 VTOL with our pixhawk VTOL.

To the KK2 system we added this: 

http://www.rcgroups.com/forums/showthread.php?t=2417656

http://www.rc-levneji.cz/user/related_files/flybarless_system_manual.pdf

It is very very simple controller but ROCK SOLID. It has a GPS only for RTL. Does not have any air speed sensors etc.

SO now with this system we just transfer the throttle from the quad to the plane with a Taranis timing and delay procedure pre programmed. It is simple one start the pusher, stop the quad in 3 seconds.

We had NO transition failures nor crashes!

Also this controller does not allow us to STALL the plane. It somehow understands (probably with the GPS)  that we are slowing and slowing and it starts cut our pitch up roll commands. When we continue to slow down and pitch up at some point the plane stalls but this little controller is able to control e very managable parachuting with just a little lest and right leaf movement. It does not go in roll spiral nothing. Just give a little throttle and it picks up from there... 

Getting in mind the KK2 control of the quad it starts to appear to me that sometimes simpler is better. And also not a bad idea to chanllenge yourself and revising your initial ideas and concepts.

BTW we did very very good PID for the quad of our VTOL with PX4 and it almost hadles like the KK. Still we have problems with the Z axis, and the quad dips on strong yaw and on sharp consequitive roll commands left and right. On strong pitch front and back it is perfect.

Surprising to me was the numbers we have got:

MC_PITCHRATE_D	0.0052
MC_PITCHRATE_FF	0.0
MC_PITCHRATE_I	0.0
MC_PITCHRATE_MAX	220.0
MC_PITCHRATE_P	0.22
MC_PITCH_P	8.28

MC_ROLLRATE_D	0.0056999996
MC_ROLLRATE_FF	0.0
MC_ROLLRATE_I	0.0
MC_ROLLRATE_MAX	220.0
MC_ROLLRATE_P	0.26
MC_ROLL_P	8.0

We used rotary knobs this time to change the params in real time. 

BTW the interface for real time param change with rotary knob in QGC should be revised.

1. We cant see what is the default assignments of param knobs to params.
2. We cannot see in real time the value we change! we have to refresh and read it from the FC after we have changed the value.

**Comment by sanderux** [comment]
@tubeme all fixed wing vehicles require trim for stable flight, whats wrong with that? 
And you KK board doesnt navigate autonomous, under those circumstances a transition is indeed VERY VERY simple.

Thanks for tuning your vehicle, a PR with the params for that model would be great!

**Comment by tubeme** [comment]
@sanderux If by trim you mean the heading orientation trim, or the reflex we put sometimes to the plane then I agree. But in my opinion in all modes other than acro modes the autopilot does the job of some sort of trimming. In the GPS guided modes the trimming the computer does is even more while keeping a certain course. So nothing wrong with the trim but I see at as manual override and and it is like an exception. So the trim in most cases is applicable to certain conditions. For a different conditions you will need different trim settings. While if the computer does its gob it will dynamically apply the necessary trimming in order to follow its dynamic characteristics.

Indeed the KK transition is very very simple. But I don't see any difference between making transition in Manual mode in PX4 and making transition from KK to the Bigaole BGL-6G-AP. Of course for the GPS guided modes the story is different. But why make a more complex the transition and implement a guided mode in it? The matter of fact is that if you start navigating during transition and apply roll and pitch it will most likely fail because the VTOL is heavier than regular quad and more inertial while in hover and during transition without lift the plane is heavy aerodynamicaly. So by presumption we have to guarantee that during transition we have a perfect level flight. On the basis of the aircraft properties we can easily calculate the length of the vector we need for transition so we need to just orient correctly the VTOL with the quad in the right direction and at the right altitude when in autonomous mission. With manual transition the orientation is done by the pilot. So in this scenarios the transition is not guided procedure which is not necessary to be complex. As you mentioned I remember you had a very good results while you experimented in the beginning letting the quad to do more stabilization during transition.

Listen the PX4 transition is the best! We have never had such a smooth transition with the parallel KK system. The problem is we have had many failures during transition with the PX4. All sorts of problems stalling, roll spins, cutting the throttle of the quad earlier, or not cutting it at all. We understand that it is in the development curve so we continue working on it. Perfecting more and more the reliability. I read in another issue by you that there is a problem with keeping altitude during transition:

https://github.com/PX4/Firmware/issues/4079

And we are working on the VTOL here in BG together for a while and we will try the transition with a master I loaded 2 weeks ago. So We are kind of afraid what is going to happen. And that is not a good sign you know. We just want 100/100 successful transitions not 70/100. For now the reason for this is ours most likely bad PIDs other probloems etc. Soon we will have the perfect PIDs and then we will report again on transitions. 

Thank you very much for dealing with the airspeed gust problems during transition.

Here are some logs.

http://logs.uaventure.com/view/2GuqDJnBQ42CvCVsFu4npd

http://logs.uaventure.com/view/emytXPBxnJKBvr5HUXC7Vf

**Comment by sanderux** [comment]
@tubeme in order to support autonomous transitions we need a lot more then simply mixing over some throttle. This indeed impacts manual transitions. We are now looking in to dealing with the altitude problem and it is likely this "fix" will not be the final solution. I can guarantee you that it is all not as simple as you make it sound but it will be fixed, just sit tight.

On your failures i have something pending that will make your flights safer, i'm working on a "parachute" function for the quad. It will basically monitor the attitude of the plane and "catch" it by enabling quad motors when things go wrong (falling, stalling, etc). That should make tuning your model safer.

**Comment by sanderux** [comment]
I had a look at your logs, your vtol seems to have some issues;
- Your airspeed sensor is not correctly calibrated or it is too close to other things like radio module. It's reading 10m/s constantly. This could be a cause of your stalls.
- Your quad seems very underpowered, i would suggest different motor/prop combo or different voltage if your drive can handle it.
- Your quad does not seem to be on the CG, this makes a lot of difference.

Look at your actuators in your logs. For reference, this is what a "normal" graph would look like on a vtol: http://logs.uaventure.com/view/djVcLjJZChq3QsUmoBsrWh#Actuators_PLOT

Trying to tune your vtol based on these issues will probably not work

**Comment by tubeme** [comment]
@sanderux Thanks a lot. We sit very tightly :) I know you guys are working on the issues and I know that it is a complex matter. But I try to challenge the concepts I'm working on all the time. It is good for the health of the project to have voices like mine.

I follow the PRs and know about the parachuting and we are waiting to test it.

**Comment by tubeme** [comment]
Thanks for the log you provided me. It is very good example. BTW I will try to place some examples in the users manual in the PID tuning guides. Because it is much more easier to explain things with graph examples.

for the issues we have:

Your airspeed sensor is not correctly calibrated or it is too close to other things like radio module. It's reading 10m/s constantly. This could be a cause of your stalls.

I know this and I've seen it but because now we are flying only the quad we did not re-calibrate our airspeed and we did not pay attention to its readings. I saw this in the graphs though. Thanks

Your quad seems very under-powered, i would suggest different motor/prop combo or different voltage if your drive can handle it.

We are using a classical config for the combo tested simonK ESC (Afro) and tested and robust NTM motors, we use BLheli as well but this logs i sent were with SimonK. Truly tested combo in all conditions. With the KK we have very adequate throttle responses with this same combo. We can shoot very very fast climbs. Somehow with the PX4 we don't have this response. The throttle response is lagy, especially if the plane starts dipping, when we apply full throttle to compensate for the dip, it starts with 0.5-1 second delay until it applies the throttle. This way the quad recovers but it looses about 3-4-5 meters before it does so. Probably the Logs I gave are from a little drained battery. But the same problem is occurring when the batteries are full just it does it a little less amplitude. When the batteries drain the problem becomes more and more obvious.

Actually the problem is not that we don,t have a good throttle. It is like the throttle is on alcohol :) it responds with delay in Auto and in Manual. And pretty obvious delay.

Your quad does not seem to be on the CG, this makes a lot of difference.

The Quad is perfectly on the CG. We discovered this rule 4 years ago. So we are very cautious double checking this. In our case we had 9.5 degrees Y offset to the heading of the FC (it is how it is mounted now) and 0.5 degrees offset to X as well. It is set by multiple measurements with level tools including many event horizon calibrations. Final tuning was done in the air with rotary knobs on the X and Y sens offset until the quad stays perfectly still in the air in manual. So you have to apply some offset in flight plot to see the real tracking.

Thanks again a lot for helping me with this.

**Comment by LorenzMeier** [comment]
@tumbili Can you review / comment here?

File: src/modules/vtol_att_control/standard.cpp [hunk 1]
```diff
@@ -309,12 +309,16 @@ void Standard::fill_actuator_outputs()
 
 	/* fixed wing controls */
 	_actuators_out_1->timestamp = _actuators_fw_in->timestamp;
-	_actuators_out_1->control[actuator_controls_s::INDEX_ROLL] = -_actuators_fw_in->control[actuator_controls_s::INDEX_ROLL]
-			* (1 - _mc_roll_weight);	//roll
+
+	_actuators_out_1->control[actuator_controls_s::INDEX_ROLL] =
+		(-_actuators_fw_in->control[actuator_controls_s::INDEX_ROLL] + _params->fw_roll_trim);
```
Comment (SimonWilks) [inline src/modules/vtol_att_control/standard.cpp]: I feel this is treating symptoms and we should still try to understand underlying causes before adding more parameters.

Are you sure you don't just have a fixed-wing tuning issue? If adding a constant trim value here is able to level it out that would suggest that means the controller should be able to do it itself if correctly tuned.

If it is only during the initial transition period that you lack roll/pitch authority have you tried increasing the blending threshold so it doesn't start reducing MC input until a higher airspeed is reached? If you feel MC control isn't powerful enough to correct during transition (is your setup still underpowered?) perhaps we need to address how blending works (ie: nonlinear; full FW control surface when transition starts; decouple MC/FW actuator blending).

Also wrt pitch is your mechanical setup such that zero pitch in MC mode is also zero pitch for FW?

This change also results in the control surfaces always being active in MC mode (needs testing in wind). So on one hand the control surfaces will now still move while in MC mode and there will also be offsets (trims) present which are intended to offset motor torque that isn't present in MC mode. Holding position in a headwind in MC mode with off center ailerons for example would result in inducing roll.

**Comment by sanderux** [comment]
Ada and i are working on the underlying problem. The issue is that the MV attitude setpoints are published and there seems to be problem with pitch and tecs.

The issue also exists in sitl and is very consistent. It does not seem to be mechanical or airframe specific.

Seting the trims is indeed fighting symptoms but the underlying problem is quite complex, this will allow a temp solution untill we have addressed the other issues.

The pitch trim param is already there, having a roll trim sounds logical from that point of view. It could also be used for quadplanes that have a different sense of level for quad and plane flight.

**Comment by tubeme** [comment]
Working on many projects I learned something from my experience. It is never good practice to implement temporary patches during development. Better loose some time dealing fundamentally with the problem and solve it once and for all, than patch work. Patch work usually generates problems at other places and the final verdict is - you loose more time finally to deal with not one but many more generated problems.

I will gladly wait a 1 month or even 2 months for the right solution patiently if I know it will disappear.

Reading Simon Wilks comments I realized we've made a FUNDAMENTAL mistake.

BY this:
**
If it is only during the initial transition period that you lack roll/pitch authority have you tried increasing the blending threshold so it doesn't start reducing MC input until a higher airspeed is reached? If you feel MC control isn't powerful enough to correct during transition (is your setup still underpowered?) perhaps we need to address how blending works (ie: nonlinear; full FW control surface when transition starts; decouple MC/FW actuator blending).**

I understand that the beyond blending speed threshold, the MC is diminishing its control over the craft in some sort of fade off. 

If this is correct we made a huge mistake. Our blending speed is 2 m/s because we want elevons control right away. In this case as soon as possible. This way the quad diminishes its control very soon and the plane levons does not play huge role in stabilization until lift is produced. This way we had weightless machine during transition. So perfect transition in dead weather and problems in strong weather.

Now there is very important detail about the VTOLs in general that I will share with you. During hover mode we want both stabilization to work! The quad should does its job of main stabilization, but at the same time to have a full elevon or other control surfaces to move and do their work like in forward flight. Something like the permanent stabilization in forward flight but all the time even in hover flight.

This gives tremendous stabilization in strong weather. because if the wind is about 5 m/s the gusts produce lift and then the elevons start helping the quad in stabilization. Given that the VTOL is not the best flying MC, given that in most cases it is under-powered quad it is very important to allow both aircraft to do the stabilization simultaneously. The quad is under powered because we want performance in term of time of flight and we have to put the lightest possible combo. We have a huge penalty for the weight of motors and esc's

So if we have a full totally permanent FW stabilization and quad stabilization during the whole transition period we will produce perfect results. If both work to stabilize the craft during transition and then after trans airspeed threshold just fade out the quad for a 1-1.5 seconds will be the best for us. The timing I give is an example don't get on it.

As I told you we are flying our VTOL for many years. We experimented almost anything. To give you an example with our simple KK system. Because with KK and manual direct drive plane you can turn on both throttles we had flown the Siniger with quad on 50% throttle ie hovering in the air and then with the pusher motor controlling the forward flight. The plane flew very very interesting but stable in all means. Like we had put a gyro on the plane. It just felt like ice skating making very slippery turns but in full control and the most important level flight. The moment you release the pitch/roll it just stabilized perfect and this was done by the quad because we did not have FC for the FW flight.

So the conclusion is that if physically things are OK like the CG to be at the same spot for both, both vehicles should behave and fly with almost the same characteristic in terms of stabilization, both vehicles should not have any trims especially a difference in the alpha during flight - stabilizing both at the same time is not a problem even it brings better results.

**Comment by sanderux** [comment]
@tubeme this should solve it: https://github.com/PX4/Firmware/issues/4091

**Pull request closed without merging**